### PR TITLE
PT-1293 add rescueCarbs prescriptors

### PR DESCRIPTION
**Pull request by franck-fourel**
Add 2 new optional fields for food of type 'rescuecarbs'
- prescriptor
- prescribedNutrition

The prescriptor is mandatory if prescribedNutrition is there.

### CODE REVIEW

File: data/types/food/food.go [hunk 1]
```diff
@@ -97,6 +112,19 @@ func (f *Food) Validate(validator structure.Validator) {
...
+			validator.String("prescriptor", f.Prescriptor).Exists()
+		}
+		if f.Prescriptor != nil {
+			validator.String("prescriptor", f.Prescriptor).OneOf(Presciptors()...)
+		}
+	} else {
+		// Prescriptors are ignored for other types of meals
+		f.Prescriptor = nil
+		f.PrescribedNutrition = nil
+	}
```
Comment (CedricPerotto) [inline data/types/food/food.go]: Updating data in the Validate function is a bit strange to me this step should be done in Parse or Normalize step (I would place it in Normalize, even if the order of operation seems to be Parse->Validate->Normalize).

File: data/types/food/food.go [hunk 2]
```diff
@@ -97,6 +112,19 @@ func (f *Food) Validate(validator structure.Validator) {
 	if f.Nutrition != nil {
 		f.Nutrition.Validate(validator.WithReference("nutrition"))
 	}
+	if f.Meal != nil && *f.Meal == MealRescueCarbs {
+		if f.PrescribedNutrition != nil {
+			f.PrescribedNutrition.Validate(validator.WithReference("nutrition"))
+			validator.String("prescriptor", f.Prescriptor).Exists()
```
Comment (CedricPerotto) [inline data/types/food/food.go]: According to the spec, prescriptor field is optional, and prescribedNutrition is required only if prescriptor = 'hybrid'.
Not sure if we should change the spec (for devDroids implementation) or stick to what we wrote.

**Review by CedricPerotto — CHANGES_REQUESTED** [review]

File: data/types/food/food.go (hunk 1 above)
Comment (franck-fourel) [reply data/types/food/food.go]: correct. 
moved to Normalize even if it breaks all the tests ;-)

File: data/types/food/food.go (hunk 2 above)
Comment (franck-fourel) [reply data/types/food/food.go]: updated to stick to the spec.

**Comment by franck-fourel** [comment]
the prescriptor type is redefined in https://github.com/mdblp/platform/pull/50 so that it can be shared across types

File: data/types/food/food.go [hunk 3]
```diff
@@ -97,6 +113,19 @@ func (f *Food) Validate(validator structure.Validator) {
...
 	}
+	if f.Meal != nil && *f.Meal == MealRescueCarbs {
+		if f.Prescriptor != nil {
+			validator.String("prescriptor", f.Prescriptor).OneOf(Presciptors()...)
+		}
+
+		if *f.Prescriptor == HybridPrescriptor && f.PrescribedNutrition == nil {
+			// Prescribed Nutrition is mandatory
+			validator.WithReference("prescribedNutrition").ReportError(structureValidator.ErrorValueNotExists())
+		}
```
Comment (CedricPerotto) [inline data/types/food/food.go]: Sending rescuecarbs meal with a nil prescriptor raises a 500 Error. Due to this test `*f.Prescriptor == HybridPrescriptor`.

```suggestion
		if f.Prescriptor != nil {
			validator.String("prescriptor", f.Prescriptor).OneOf(Presciptors()...)
			if *f.Prescriptor == HybridPrescriptor && f.PrescribedNutrition == nil {
				// Prescribed Nutrition is mandatory
				validator.WithReference("prescribedNutrition").ReportError(structureValidator.ErrorValueNotExists())
			}
		}
```

Comment (franck-fourel) [reply data/types/food/food.go]: applied in 024d276

**Review by yvon-dblg — COMMENTED** [review]
Documentation:
- In the documentation `docs/UpdatedData.md`, § food: The "meal" item is present 2x. JSON spec do not allow this.
- In the examples: Remove timezoneOffset, and add only timezone informations.

**Review by CedricPerotto — CHANGES_REQUESTED** [review]
Following [PR#50](https://github.com/mdblp/platform/pull/50) review:
github.com/tidepool-org/platform/data/types/bolus/prescriptor should be moved to github.com/tidepool-org/platform/data/types/common/prescriptor

File: docs/UpdatedData.md [hunk 4]
```diff
@@ -8,6 +8,8 @@ Here we document the data types that have to be created or udpated with new fiel
 - deviceEvent - Zen mode
 - deviceEvent - Private Mode
 
+_Note_: the examples below focused on the new fields. All the other fields (such as time, timezone, timezoneOffset) are not impacted by those changes and will not reauire updates.
```
Comment (CedricPerotto) [inline docs/UpdatedData.md]:
```suggestion
_Note_: the examples below focused on the new fields. All the other fields (such as time, timezone, timezoneOffset) are not impacted by those changes and will not require updates.
```

**Review by CedricPerotto — CHANGES_REQUESTED** [review]
Just a typo in doc. Apart from that LGTM. @yvon-dblg  ?

File: data/types/food/food.go [hunk 5]
```diff
@@ -97,6 +104,18 @@ func (f *Food) Validate(validator structure.Validator) {
 	if f.Nutrition != nil {
 		f.Nutrition.Validate(validator.WithReference("nutrition"))
 	}
+	if f.Meal != nil && *f.Meal == MealRescueCarbs {
+		if f.Prescriptor != nil {
+			f.Prescriptor.Validate(validator)
+			if *f.Prescriptor.Prescriptor == common.HybridPrescriptor && f.PrescribedNutrition == nil {
+				// Prescribed Nutrition is mandatory
```
Comment (yvon-dblg) [inline data/types/food/food.go:111]:
```
auto: nutrition and prescribedNutrition are equal
```
When prescriptor is auto we can send different prescribedNutrition  & nutrition values.
Should a warning or an error issued ?
Maybe remove the prescribedNutrition  ?

Comment (yvon-dblg) [reply data/types/food/food.go:111]:
```
manual: prescribedNutrition is ignored
```
Should we remove the prescribedNutrition in manual prescriptor ?

**Review by yvon-dblg — COMMENTED** [review]
2 questions.
Plus the previous comment made.

File: data/types/food/food.go (hunk 5 above)
Comment (franck-fourel) [reply data/types/food/food.go:111]: makes sense to ignore the prescribedNutrition for auto and manual

**Review by yvon-dblg — APPROVED** [review]
LGTM

**Review by CedricPerotto — APPROVED** [review]

**Pull request merged**